### DISCUSSION TITLE
Failing to save content with the Elements.io sample

### DISCUSSION THREAD

**Discussion by eduardoarantes** [Hieromon/AutoConnect#579] [Q&A] [answered]
I am using a NodeMCU board and playing with the Elements.io sample

I had to modify the code slightly because it wasn't compiling. The changes were only on the headers import and they look like this:
The rest of the code remains exactly like https://github.com/Hieromon/AutoConnect/blob/master/examples/Elements/Elements.ino

`#if defined(ARDUINO_ARCH_ESP8266)
#include <ESP8266WiFi.h>
#include <ESP8266WebServer.h>
#include <ESP8266HTTPClient.h>
#include <ESP8266mDNS.h>
#define FORMAT_ON_FAIL
using WebServerClass = ESP8266WebServer;
#elif defined(ARDUINO_ARCH_ESP32)
#include <WiFi.h>
#include <WebServer.h>
#include <HTTPClient.h>
#include <ESPmDNS.h>
#define FORMAT_ON_FAIL  true
using WebServerClass = WebServer;
#endif

#include <AutoConnect.h>

#include <LittleFS.h>
FS& FlashFS = LittleFS;`

The execution prints "Filesystem failed to open."

How do I troubleshoot this? Is there a way to collect more detailed information from the LittleFS library about the problem?

My platformio.ino looks like this:

`
[env:nodemcuv2]
platform = espressif8266
board = nodemcuv2
framework = arduino
lib_ldf_mode = deep+
monitor_speed = 115200
monitor_filters = direct
lib_deps = 
	hieromon/PageBuilder@^1.5.6
	hieromon/AutoConnect@^1.4.2
build_flags =
    -DAC_USE_SPIFFS
    -DPB_USE_SPIFFS
board_build.filesystem = littlefs
`

tks

**Comment by Hieromon** [comment, marked as answer]
Your specified `board_build.filesystem=littlefs` and filesystem identifiers for AutoConnect, and PageBuilder are different by `build_flags = -DAC_USE_SPIFFS -DPB_USE_SPIFFS`. It tells AutoConnect and PageBuilder to use SPIFFS.

The correct file system identifier for NodeMCUV2 with LittleFS is as follows:
```ini
build_flags = -DAC_USE_LITTLEFS -DPB_USE_LITTLEFS
```

See: https://hieromon.github.io/AutoConnect/filesystem.html

**Reply by eduardoarantes** [reply]
@Hieromon ,
thanks.
It kind of works, but it doesn't persist after resetting the board.
Is there a missing parameter?

**Reply by Hieromon** [reply]
@eduardoarantes 
Ok, I will try to diagnose the problem on my end. Please rebuild with the following `platformio.ini`. That way can get a debug trace with [**AC_DEBUG**](https://hieromon.github.io/AutoConnect/adothers.html#debug-print) enabled:

```ini
framework = arduino
platform = espressif8266
board = nodemcuv2
board_build.filesystem = littlefs
lib_ldf_mode = deep
lib_deps =
	hieromon/PageBuilder@^1.5.6
	hieromon/AutoConnect@^1.4.2
build_flags =
	-DAC_DEBUG
	-DAC_USE_LITTLEFS
	-DPB_USE_LITTLEFS
upload_speed = 921600
monitor_speed = 115200
monitor_filters = esp8266_exception_decoder
```

Then, please post the following output. I will make a diagnosis based on the results.

- The output log of Dependency graph at build time with PlatformIO build
- Serial output log including AC_DEBUG trace during sketch execution

**Reply by eduardoarantes** [reply]
I couldn't set `upload_speed = 921600`. It didn't upload the sketch

`Scanning dependencies...
Dependency Graph
|-- PageBuilder @ 1.5.6
|   |-- ESP8266WiFi @ 1.0
|   |-- ESP8266WebServer @ 1.0
|   |-- LittleFS @ 0.1.0
|-- AutoConnect @ 1.4.2
|   |-- PageBuilder @ 1.5.6
|   |   |-- ESP8266WiFi @ 1.0
|   |   |-- ESP8266WebServer @ 1.0
|   |   |-- LittleFS @ 0.1.0
|   |-- ArduinoJson @ 6.20.1
|   |-- ESP8266WiFi @ 1.0
|   |-- ESP8266WebServer @ 1.0
|   |-- DNSServer @ 1.1.1
|   |   |-- ESP8266WiFi @ 1.0
|   |-- EEPROM @ 1.0
|   |-- LittleFS @ 0.1.0
|   |-- SD @ 2.0.0
|   |   |-- SDFS @ 0.1.0
|   |   |   |-- SPI @ 1.0
|   |   |   |-- ESP8266SdFat @ 2.1.1
|   |   |   |   |-- SPI @ 1.0
|   |   |-- SPI @ 1.0
|   |   |-- ESP8266SdFat @ 2.1.1
|   |   |   |-- SPI @ 1.0
|   |-- SDFS @ 0.1.0
|   |   |-- SPI @ 1.0
|   |   |-- ESP8266SdFat @ 2.1.1
|   |   |   |-- SPI @ 1.0
|   |-- SPI @ 1.0
|   |-- ESP8266SdFat @ 2.1.1
|   |   |-- SPI @ 1.0
|   |-- Ticker @ 1.0
|   |-- ESP8266HTTPClient @ 1.2
|   |   |-- ESP8266WiFi @ 1.0
|   |-- ESP8266httpUpdate @ 1.3
|   |   |-- ESP8266WiFi @ 1.0
|   |   |-- ESP8266HTTPClient @ 1.2
|   |   |   |-- ESP8266WiFi @ 1.0
|-- ESP8266WiFi @ 1.0
|-- ESP8266WebServer @ 1.0
|-- ESP8266HTTPClient @ 1.2
|   |-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
|   |-- ESP8266WiFi @ 1.0
|-- DNSServer @ 1.1.1
|   |-- ESP8266WiFi @ 1.0
|-- EEPROM @ 1.0
|-- LittleFS @ 0.1.0
|-- SD @ 2.0.0
|   |-- SDFS @ 0.1.0
|   |   |-- SPI @ 1.0
|   |   |-- ESP8266SdFat @ 2.1.1
|   |   |   |-- SPI @ 1.0
|   |-- SPI @ 1.0
|   |-- ESP8266SdFat @ 2.1.1
|   |   |-- SPI @ 1.0
|-- SDFS @ 0.1.0
|   |-- SPI @ 1.0
|   |-- ESP8266SdFat @ 2.1.1
|   |   |-- SPI @ 1.0
|-- SPI @ 1.0
|-- ESP8266SdFat @ 2.1.1
|   |-- SPI @ 1.0
|-- Ticker @ 1.0
|-- ArduinoJson @ 6.20.1
|-- ESP8266httpUpdate @ 1.3
|   |-- ESP8266WiFi @ 1.0
|   |-- ESP8266HTTPClient @ 1.2
|   |   |-- ESP8266WiFi @ 1.0
Building in release mode
Compiling .pio/build/nodemcuv2/src/Elements.ino.cpp.o
Retrieving maximum program size .pio/build/nodemcuv2/firmware.elf
Checking size .pio/build/nodemcuv2/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [=======   ]  70.7% (used 57916 bytes from 81920 bytes)
Flash: [=====     ]  49.8% (used 519931 bytes from 1044464 bytes)`


`[AC] Host:192.168.1.22,/elements,generated:/elements, allocated
[AC] Host:192.168.1.22,/elements,already allocated
[AC] CB in AHEAD /elements
[AC] airconIP<4> of /elements loaded
[AC] zoneId<4> of /elements loaded
[AC] Host:192.168.1.22,/save,generated:/save, allocated
[AC] Host:192.168.1.22,/save,already allocated
[AC] fetch /elements, elements stored
[AC] CB in AHEAD /save
[AC] JSON buffer size:784
`

**Reply by Hieromon** [reply]
I don't see any signs of a problem. But there are too few `[AC]` logs. Have you truncated some of them?

>  it doesn't persist after resetting the board.

What phenomenon is your comment above specifically? Could it be that you are running out of memory?
The ESP8266 Arduino core has increased memory consumption since 3.1.0. This may cause some sketches to run out of memory. It can be determined, for example, by the following `[AC]` log:

```
[AC] Deserialize:NoMemory
```

See AutoConnect v1.4.2 [Release note](https://github.com/Hieromon/AutoConnect/releases/tag/v1.4.2).

**Reply by eduardoarantes** [reply]
@Hieromon 
I when trying to parse the json content of the saved file I got the "JSON de-serialization failed: NoMemory".

That is the reason.

I will read the Release note

**Reply by eduardoarantes** [reply]
@Hieromon ,

I realised the Out of memory error was when I was parsing the json.
I fixed that, and I can see the value has been stored.

It's just not re populating that page when it loads after restarting
